More /vg/lighting tweaks

MrsTonedOne opened the pull request:
Rather than remove the light from all tiles, then re-add the light to all tiles, we just go thru each tile and diff the light level from the last value we added to it. (since this is tracked)
This cut the proc calls for updating light sources on movements in half.

Lighting objects now default to full brite rather than full dark so shuttles aren't as immersion breaking when they transit.

Made lighting more aggressive about clearing empty lists. (Memory usage went down ~80mb)

Break all lights followed by fix all lights went down from 3.5s cpu time to 2.4s cpu time.

Because this is a decent sized change in how lighting_source updates operate, its hard to get a true apples to apples comparison by comparing two procs like i normally would.

:cl:
tweak: Lighting now defaults to fully bright until the first update tick for that tile. This makes shuttle movements less immersion breaking.
experiment: Rather than remove the light source from the system and then re-adding it on light movements, the system now calculates and applies the difference to each tile. This should speed up lighting updates.
/:cl:

@PJB3005 i merged smart_vis_update and apply_lum and check() into one proc by making APPLY_CORNER() diff the old value (if any) from the new value.

@Incoming5643 this should fix tiles that started out dark not getting the dark icon state

MrsTonedOne committed:
More /vg/lighting tweaks
Rather then remove the light from all tiles, then re-add the light to all tiles, we just go thru each tile and diff the light level from the last value we added to it. (since this is tracked)
This cut the proc calls for updating lights in half.

Lighting objects now default to full brite rather then full dark so shuttles aren't as immersion breaking when they transit.

Made lighting more agressive about clearing empty lists.

RemieRichards commented on code/modules/lighting/lighting_object.dm:
```diff
@@ -69,10 +68,17 @@ GLOBAL_LIST_EMPTY(all_lighting_objects) // Global list of lighting objects.
+	var/datum/lighting_corner/cb = dummy_lighting_corner
+	var/datum/lighting_corner/ca = dummy_lighting_corner
+	if (corners) //done this way for speed
+		cr = T.corners[3] || dummy_lighting_corner
```
could you not then accessing these from ```corners```, the local should be faster

MrStonedOne replied:
oh ya

MrsTonedOne committed:
Does the remie

MrsTonedOne committed:
Missed a spot
Lighting_corners should now nolonger have an init proc

MrPerson commented on code/modules/lighting/lighting_source.dm:
```diff
@@ -164,48 +110,47 @@
+	. *= light_power;                        \
+	var/OLD = effect_str[C];                 \
+                                             \
+	effect_str[C] = .;                       \
```
you could do this on one line with effect_str[C] = . = LUM_FALLOFF(C, pixel_turf) * light_power although that might be a too much

MrsTonedOne committed:
Some more improvments, removed dview.

This *seems* to be a decent speed up just in the appearent time it takes break all lights to process, and dview wasn't really needed here.

Minor oddies can occur on bigger lights near their edges, but not enough to justify the extra cost.

ChangelingRain commented:
Yeah space doesn't light walls, it only emits light next to open turfs

ChangelingRain commented:
From that issue, I can say emitter beams also cause that.

octareenroon91 commented:
See those dark spots? I only grew and removed glowshrooms. Zero shadowshrooms.

There was also a weird shadowy spot in the arrival shuttle before this round.

![screenshot_2017-05-06_22-32-00](https://cloud.githubusercontent.com/assets/12420976/25777765/0fbc1aa6-32ad-11e7-8db1-bb1c470a051b.png)
![screenshot_2017-05-06_22-35-54](https://cloud.githubusercontent.com/assets/12420976/25777766/0fbd0b8c-32ad-11e7-9624-ae692ea88ea2.png)

Values saturate but subtraction doesn't know.

Lohikar commented on code/modules/lighting/lighting_source.dm:
```diff
@@ -164,48 +110,47 @@
+	(                                        \
+		(. * lum_r) - (OLD * applied_lum_r), \
+		(. * lum_g) - (OLD * applied_lum_g), \
+		(. * lum_g) - (OLD * applied_lum_g)  \
```
Shouldn't this last line be `(. * lum_b) - (OLD * applied_lum_b)`?

MrStonedOne replied:
yes

octareenroon91 commented:
![screenshot_2017-05-06_23-23-07](https://cloud.githubusercontent.com/assets/12420976/25777912/5809c5be-32b3-11e7-9204-4e0b9081d9cd.png)

MrsTonedOne committed:
Copy/paste fail

SpinnerMaster commented:
Can confirm arrivals shuttle lighting is fucked.

![image](https://cloud.githubusercontent.com/assets/4211448/25777969/8bcbf6cc-32b5-11e7-89ac-d82823466b40.png)

MrsTonedOne committed:
Fix starlight

MrStonedOne commented:
all the issues seem to be fixed.

AnturK commented:
![](http://i.imgur.com/iVDWuUv.png)

Still happening ? Or was the testmerge not updated ?

MrStonedOne commented:
hmmm

MrStonedOne commented:
the test merge was fixed, so this would be it still happening.

octareenroon91 commented:
>Rather than remove the light from all tiles, then re-add the light to all tiles, we just go thru each tile and diff the light level from the last value we added to it. (since this is tracked)

This change is sensitive to inconsistency in light creation/removal.

Also I'm assuming that light levels are in boundaries (-infinity, 1] so lights that don't clean up after themselves in the sanest ways will pull a turf's light level down for the rest of the round.

Consider adding occasional full recalculations.

PJB3005 commented:
@octareenroon91 light levels have no bounds other than the limits of a single precision float. They get capped in the calculation for the strength of a lighting overlay and `turf.get_lumcount()` though.

MrStonedOne commented:
>Consider adding occasional full recalculations.

So, I'm going to point out the irony that the current bug antur mentioned, is actually from a full recalculation that happens when turfs change.

MrsTonedOne committed:
Fixes light/lum counts getting fucked up when turfs changed

MrsTonedOne committed:
Fixes a bug in processing lights of deleted atoms.

MrsTonedOne committed:
Clearly the coffee hasn't kicked in yet

MrsTonedOne committed:
last one I swear mom!

optimumtact commented:
those bugs look hilarious haha, we should have some kind of light that can achieve that effect.

ChangelingRain commented:
![](http://puu.sh/vIVvy/322d3fa8fe.png)
Still SOME issues.

Razharas commented:
Cut and build instead of readjusting is a method plaguing our codebase in all systems
We need more of this

PJB3005 commented:
I was totally thinking of being more efficient with the moving of light sources, but it just gave me a giant headache to think about and all the edge cases and I gave up.

Vis updates do only update what needs updating already though.

MrsTonedOne committed:
>forgetting that REMOVE_EFFECT needs effect_str

MrsTonedOne committed:
Fixes nested lights not working properly
